addons/mail/models/mail_canned_response.py

```diff
@@ -0,0 +1,88 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models, api
+
+
+class MailCannedResponse(models.Model):
+    """
+    Canned Response: content that will automatically replace the shortcut of your choosing. This content can still be adapted before sending your message.
+    """
+
+    _name = "mail.canned.response"
+    _description = "Canned Response"
+    _order = "id desc"
+    _rec_name = "source"
+
+    source = fields.Char(
+        "Shortcut", required=True, index="trigram",
+        help="Canned response that will automatically be substituted with longer content in your messages."
+        " Type ':' followed by the name of your shortcut (e.g. :hello) to use in your messages.",
+    )
+    substitution = fields.Text(
+        "Substitution",
+        required=True,
+        help="Content that will automatically replace the shortcut of your choosing. This content can still be adapted before sending your message.",
+    )
+    description = fields.Char("Description")
+    last_used = fields.Datetime("Last Used", help="Last time this canned_response was used")
+    group_ids = fields.Many2many("res.groups", string="Authorized Groups", groups="mail.group_mail_canned_response_admin")
+    is_shared = fields.Boolean(
+        string="Determines if the canned_response is currently shared with other users",
+        compute="_compute_is_shared",
+        store=True,
+    )
+    is_editable = fields.Boolean(
+        string="Determines if the canned response can be edited by the current user",
+        compute="_compute_is_editable"
+    )
+
+    @api.depends("group_ids")
+    def _compute_is_shared(self):
+        for canned_response in self:
+            # sudo: regular user should be able to compute
+            canned_response.is_shared = bool(canned_response.sudo().group_ids)
+
+    @api.depends_context('uid')
+    @api.depends("create_uid")
+    def _compute_is_editable(self):
+        for canned_response in self:
+            if not canned_response.id:
+                canned_response.is_editable = True
+            else:
+                try:
+                    canned_response.check_access_rights("write")
+                    canned_response.check_access_rule("write")
+                    canned_response.is_editable = True
+                except: # pylint: disable=bare-except
+                    canned_response.is_editable = False
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        for canned_response in res:
+            payload = {"CannedResponse": canned_response._read_format(["id", "source", "substitution"])}
+            notifications = [(self.env.user.partner_id, "mail.record/insert", payload)]
+            # sudo: avoid an access error for user creating a personnal canned responses
+            for group in canned_response.sudo().group_ids:
```

Comment:
Why a sudo ? Internal user have read access on those, isn't it ? In which case, valid for other sudo usage :)

Reply:
`group_ids` is a restricted field, (see the field definition) but it should not stop the notification flow from proceeding.

Reply:
It is strange to restrict it, as you should be able to read its value. I understand you don't want people to modify it unless you are an admin to avoid letting people bloat the UI... I would do it the other way round (not mandatory, just more inlined with what we do generally I think)
  * remove the "groups" on the field -> add it in the view instead so that it is not available for standard users;
  * override create / write and either pop 'groups_id' from vals, either raise (would pop the value personnally, don't think we have to raise just for that);
  * remove sudo as now people can read the field normally;

You have to override create / write (but it is already done), but you gain an easier read access to the model, don't have to sudo the reading of groups, ... so it relies more on standard ACLs and 2many fields access.

Reply:
Unless I'm mistaken, but `groups' feel like a standard ACL feature?

I don't like your solution because the security aspect rely on custom code in override. I really don't want to write any security related code myself.

It's also more line of code to avoid some harmless `sudo`.

Reply:
ACL means access control list, nothing to do with field access. Those are two different concept. Adding "groups" on a field is used when a field contains critical information people should not access like access_token, credentials, ... Here it is not really the case.

Reply:
You are right, but the point is that it's a standard security feature I can trust to be solid. 
The usage != the purpose. You are using it for access_token and credentials, but the purpose is the restrict a field to a user group.

Reply:
Would say this discussion depends on the other question: can I update / delete canned response that I created but that have a group on it ? A "promoted" canned response in some sort (it could happen so we have to think about it).